packager/media/formats/mp4/low_latency_segment_segmenter.cc

```diff
@@ -0,0 +1,235 @@
+}
+
+LowLatencySegmentSegmenter::~LowLatencySegmentSegmenter() {}
+
+bool LowLatencySegmentSegmenter::GetInitRange(size_t* offset, size_t* size) {
+  VLOG(1) << "LowLatencySegmentSegmenter outputs init segment: "
+          << options().output_file_name;
+  return false;
+}
+
+bool LowLatencySegmentSegmenter::GetIndexRange(size_t* offset, size_t* size) {
+  VLOG(1) << "LowLatencySegmentSegmenter does not have index range.";
+  return false;
+}
+
+std::vector<Range> LowLatencySegmentSegmenter::GetSegmentRanges() {
+  VLOG(1) << "LowLatencySegmentSegmenter does not have media segment ranges.";
+  return std::vector<Range>();
+}
+
+Status LowLatencySegmentSegmenter::DoInitialize() {
+  return WriteInitSegment();
+}
+
+Status LowLatencySegmentSegmenter::DoFinalize() {
+  // Update init segment with media duration set.
+  RETURN_IF_ERROR(WriteInitSegment());
+  SetComplete();
+  return Status::OK;
+}
+
+Status LowLatencySegmentSegmenter::DoFinalizeSegment() {
+  return FinalizeSegment();
+}
+
+Status LowLatencySegmentSegmenter::DoFinalizeChunk() {
+  if (is_initial_chunk_in_seg_) {
+    return WriteInitialChunk();
+  }
+  return WriteChunk();
+}
+
+Status LowLatencySegmentSegmenter::WriteInitSegment() {
+  DCHECK(ftyp());
+  DCHECK(moov());
+  // Generate the output file with init segment.
+  std::unique_ptr<File, FileCloser> file(
+      File::Open(options().output_file_name.c_str(), "w"));
+  if (!file) {
+    return Status(error::FILE_FAILURE,
+                  "Cannot open file for write " + options().output_file_name);
+  }
+  std::unique_ptr<BufferWriter> buffer(new BufferWriter);
+  ftyp()->Write(buffer.get());
+  moov()->Write(buffer.get());
+  return buffer->WriteToFile(file.get());
+}
+
+Status LowLatencySegmentSegmenter::WriteInitialChunk() {
+  DCHECK(sidx());
+  DCHECK(fragment_buffer());
+  DCHECK(styp_);
+  
+  DCHECK(!sidx()->references.empty());
+  // earliest_presentation_time is the earliest presentation time of any access
+  // unit in the reference stream in the first subsegment.
+  sidx()->earliest_presentation_time =
+      sidx()->references[0].earliest_presentation_time;
+
+  if (options().segment_template.empty()) {
+    // Append the segment to output file if segment template is not specified.
+    file_name_ = options().output_file_name.c_str();
+  } else {
+    file_name_ = GetSegmentName(options().segment_template,
+                               sidx()->earliest_presentation_time,
+                               num_segments_, options().bandwidth);
+  }
+
+  // Create the segment file
+  segment_file_ = File::Open(file_name_.c_str(), "a");
+  if (!segment_file_) {
+    return Status(error::FILE_FAILURE, "Cannot open segment file: " +
+                                            file_name_);
+  }
+
+  std::unique_ptr<BufferWriter> buffer(new BufferWriter());
+  std::unique_ptr<File, FileCloser> file;
+
+  // Point to the already open segment file for writing
+  file.reset(segment_file_);
```

Comment:
I was really confused about this.  Why put a persistent handle in a unique_ptr, then release it afterward?  What value is unique_ptr bringing to this?

Reply:
This is an excellent question - this area was something that I was struggling with earlier. I found the flushing of the `ThreadedIoFile` and `HttpFile` and their corresponding `IoCache` buffers to be quite confusing and problematic when trying to keep the remote file open and within scope. 

This was my dirty solution to get the LL-DASH feature working, which I arrived at after a lot of fiddling and iteration. In hindsight, the implementation is kind of a wack, Frankenstein product after exploring a lot of different options.

You are totally right. The unique_ptr is not needed with the implementation above 😀 !!!

---

** NOTE: For people reading this thread who are not familiar with files within Packager, below is an outline of how `File`, `IoCacheFile`, and `HttpFile` interact. I wrote it out while reviewing the section Joey pointed out and thought that someone else might fine it helpful.

---

The call tree that I outlined is that `File::Open(...)` calls `File::Create(...)`. 

`File::Create(...)` calls `File::CreateInternalFile(...)`, which will trigger the `HttpFile` constructor via `FileFactoryFunction factory_function(...)`.  The HttpFile is pointed to by `std::unique_ptr<File, FileCloser> internal_file`.

At this point, we have our HttpFile object, `internal_file`, within `File::Create(...)`. Then, the unique_ptr `internal_file` is transferred to a new `ThreadedIoFile` via `std::move()` and stored the private variable `internal_file_`. This `ThreadedIoFile` object is returned to whoever called `File::Create(...)`. 

So, the end result of `File::Create(...)` is a `File*` to a `ThreadedIoFile`, and this `ThreadedIoFile` contains `std::unique_ptr<File, FileCloser> internal_file_`, which points to a `HttpFile` in our use case.

In `LowLatencySegmentSegmenter`, the `File* segment_file_` points to the returned `ThreadedIoFile` object, providing access to the `HttpFile`.

---

Reply:
Should segment_file_ be a unique_ptr<File>, then, to clean it up automatically when LowLatencySegmentSegmenter is destroyed?  You can still reset() the pointer when you finalize a segment or start a new one.

Reply:
This has been addressed. :)